### PR TITLE
apiextensions/examples: add typed client

### PR DESCRIPTION
**Pull request by nikhita**
It was discussed in https://github.com/kubernetes/kubernetes/pull/52753#issuecomment-337840762 to add a typed client example to apiextensions-apiserver and remove the controller part.

For an example controller, we have [`k8s.io/sample-controller`](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/sample-controller) now.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/README.md [hunk 1]
```diff
@@ -2,18 +2,29 @@
...
+be generated using the `./hack/update-codegen.sh` script.
 
-* How to register a new custom resource (custom resource type) using a CustomResourceDefinition
-* How to create/get/list instances of your new resource type (update/delete/etc work as well but are not demonstrated)
-* How to setup a controller on resource handling create/update/delete events
+The `update-codegen` script will automatically generate the following files and
+directories:
 
-## Running
+* `pkg/apis/samplecontroller/v1alpha1/zz_generated.deepcopy.go`
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/README.md]: should we change the api name? Is it used anywhere else?

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/README.md]: copy-paste error. will change this.

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/hack/update-codegen.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
+CODEGEN_PKG=${CODEGEN_PKG:-$(cd ${SCRIPT_ROOT}; ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo ../code-generator)}
+
+if LANG=C sed --help 2>&1 | grep -q GNU; then
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/hack/update-codegen.sh]: where is SED used?

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/register.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cr
+
+const (
+	GroupName = "cr.client-go.k8s.io"
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/register.go]: Technically it's not in client-go. Worth changing? Have no strong opinion.

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/register.go]: `cr.apiextensions.k8s.io`?

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/v1/register.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+var SchemeGroupVersion = schema.GroupVersion{Group: cr.GroupName, Version: "v1"}
+
+// Resource takes an unqualified resource and returns a Group qualified GroupResource
+func Resource(resource string) schema.GroupResource {
+	return SchemeGroupVersion.WithResource(resource).GroupResource()
+}
+
+var (
+	// localSchemeBuilder and AddToScheme will stay in k8s.io/kubernetes.
+	SchemeBuilder      runtime.SchemeBuilder
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/v1/register.go]: do we need both local and public scheme builder?

Also, the comment is outdated here.

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/v1/roundtrip_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+			for i := range obj.Items {
+				c.Fuzz(&obj.Items[i])
+			}
+		},
+	}
+}
+
+// TestRoundTrip tests that the third-party kinds can be marshaled and unmarshaled correctly to/from JSON
+// without the loss of information. Moreover, deep copy is tested.
+func TestRoundTrip(t *testing.T) {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/v1/roundtrip_test.go]: Does this actually do anything? I remember we had a variant for external-only types.

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/v1/types.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +resource:path=example
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/v1/types.go]: I believe this +resource:path tag doesn't do anything.

File: staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh [hunk 7]
```diff
@@ -31,49 +30,14 @@ else
...
-  --output-package k8s.io/apiextensions-apiserver/pkg/client/informers
-  "$@"
+# generate the code with:
+# --output-base    because this script should also be able to run inside the vendor dir of
+#                  k8s.io/kubernetes. The output-base is needed for the generators to output into the vendor dir
+#                  instead of the $GOPATH directly. For normal projects this can be dropped.
+${CODEGEN_PKG}/generate-groups.sh all \
+  k8s.io/apiextensions-apiserver/pkg/client k8s.io/apiextensions-apiserver/pkg/apis \
+  apiextensions:v1beta1 \
+  --output-base "$(dirname ${BASH_SOURCE})/../../.."
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh]: 👍

File: staging/src/k8s.io/sample-controller/hack/update-codegen.sh [hunk 8]
```diff
@@ -21,11 +21,20 @@ set -o pipefail
 SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
 CODEGEN_PKG=${CODEGEN_PKG:-$(cd ${SCRIPT_ROOT}; ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo ../code-generator)}
 
+if LANG=C sed --help 2>&1 | grep -q GNU; then
```
Comment (sttts) [inline staging/src/k8s.io/sample-controller/hack/update-codegen.sh]: where is SED needed?

Comment (nikhita) [reply staging/src/k8s.io/sample-controller/hack/update-codegen.sh]: will remove.

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/register.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/register.go]: cr.example.apiextensions.k8s.io? What is the CRD group? should match that.

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/register.go]: CRD group is `apiextensions.k8s.io`. cr.example.apiextensions.k8s.io seems good! :+1:

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/register.go]: :+1:

**Comment by nikhita** [comment]
All green now. PTAL, thanks.

**Pull request merged**